sql/src/main/java/io/crate/operation/udf/UserDefinedFunctionService.java

```diff
@@ -0,0 +1,239 @@
+import io.crate.settings.CrateSetting;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.cluster.AckedClusterStateUpdateTask;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateRequest;
+import org.elasticsearch.cluster.ack.ClusterStateUpdateResponse;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static io.crate.operation.udf.UserDefinedFunctionsMetaData.PROTO;
+
+@Singleton
+public class UserDefinedFunctionService {
+
+    static {
+        // register non plugin custom metadata
+        MetaData.registerPrototype(UserDefinedFunctionsMetaData.TYPE, PROTO);
+    }
+
+    public static final CrateSetting<Boolean> UDF_SETTING = CrateSetting.of(Setting.boolSetting(
+        "udf.enabled", false,
+        Setting.Property.NodeScope), DataTypes.BOOLEAN);
+
+    private final Settings settings;
+    private final ClusterService clusterService;
+    private Map<String, UDFLanguage> languageRegistry = new HashMap<>();
+
+    private static final Logger LOGGER = Loggers.getLogger(UserDefinedFunctionService.class);
+
+    @Inject
+    public UserDefinedFunctionService(Settings settings, ClusterService clusterService) {
+        this.settings = settings;
+        this.clusterService = clusterService;
+    }
+
+    public UDFLanguage getLanguage(String languageName) throws IllegalArgumentException {
+        UDFLanguage lang = languageRegistry.get(languageName);
+        if (lang == null) {
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH, "'%s' is not a valid UDF language", languageName));
+        }
+        return lang;
+    }
+
+    private boolean enabled() {
+        return UDF_SETTING.setting().get(settings);
+    }
+
+    public void registerLanguage(UDFLanguage language) {
+        if (enabled()) {
+            languageRegistry.put(language.name(), language);
+        } else if (LOGGER.isInfoEnabled()) {
+            LOGGER.info("Could not register language '{}' because UDF is disabled.", language.name());
+        }
+    }
+
+    /**
+     * This method can only be called on master node
+     * registers new function in the cluster
+     *
+     * @param request function to register
+     */
+    void registerFunction(final RegisterUserDefinedFunctionRequest request,
```

Comment:
why are the request classes here necessary? Seems like it's possible to pass the parameters directly without wrapping them.